_posts/gRPC-Go performance improvements.md

```diff
@@ -0,0 +1,148 @@
+# gRPC-Go performance improvements
+
+For past few months we've been working on improving gRPC-Go performance.This includes improving network utilization, optimizing CPU usage and memory allocations. Most of our recent effort has been focused around revamping gRPC-go flow control. After several optimizations and new features we've been able to improve quite significantly, especially on high-latency networks. We expect users that are working with high-latency networks and large messages to see an order of magnitude performance gain. 
+Please refer to some benchmark results at the end.
+
+This blog is a summary of all that we have done so far (in chronological order) to improve performance and our near-future plans.
+
+## Recently Implemented Optimizations
+
+
+### Expanding stream window on receiving large messages 
+
+[Code link](https://github.com/grpc/grpc-go/pull/1248)
+
+This is an optimization used by gRPC-C to achieve performance benefits for large messages. The idea is that when there's an active read by the application on the receive side, we can effectively bypass stream-level flow control to request the whole message. This proves to be very helpful when the message is big. Since the application is already committed to reading and has allocated enough memory for it, it makes sense that we send a proactive large window update (if necessary) to get the whole message rather than receiving it in chunks and sending window updates when we run low on window.
+
+This optimization alone provided a 10x improvement for large messages on high-latency networks.
+
+
+### Decoupling application reads from connection flow control
+
+[Code link](https://github.com/grpc/grpc-go/pull/1265)
+
+After having several discussions with gRPC-Java and gRPC-C core team we realized that gRPC-Go's connection-level flow control was overly restrictive in the sense that window updates on the connection depended upon if the application had read data from it or not. It must be noted that it makes perfect sense to have stream-level flow control depended on application read but not so much for connection-level flow control. The rationale is as floows: A connection is shared by several streams (RPCs). Now, if there were at least one stream that read slowly or didn't read at all, it would hamper the performance or completely stall other streams on that connection. This happens because we won't send out window updates on the connection until that slow or inactive stream read data. Therefore, it makes sense to decouple the connection's flow control from application reads. 
```

Comment:
s/depended/depend 
s/floows/follows

> Now, if there were at least one stream that read slowly or didn't read at all,


If there is one stream that is read slowly or not at all, it will impact the performance for all streams in the connection. This happens because window updates are not sent on the connection until the slow or inactive stream's data is read. 
Decoupling the connection's flow control from application reads removes this problem.